### ISSUE TITLE
Bug: Can't set schema in createMachine

### ISSUE THREAD

**Issue by davydof** [statelyai/xstate#3137]
### Description

[Here](https://github.com/statelyai/xstate/blob/main/packages/core/src/Machine.ts#L135) `TStateSchema` is `any` instead of generic as [here](https://github.com/statelyai/xstate/blob/main/packages/core/src/Machine.ts#L58)

Why?

### Expected result

```ts
import { createMachine } from 'xstate';

createMachine<IContext, ISchema, IEvent>(config);
```

### Actual result

```ts
import { Machine } from 'xstate';

Machine<IContext, ISchema, IEvent>(config);
```
is ok but next example is failed:

```ts
import { createMachine } from 'xstate';

createMachine<IContext, ISchema, IEvent>(config);
```
Type error: Type 'ISchema' does not satisfy the constraint 'EventObject'.
  Property 'type' is missing in type 'ISchema' but required in type 'EventObject'.  TS2344

### Reproduction

https://codesandbox.io/s/admiring-shtern-47um0b

**Comment by Andarist** [comment]
`TStateSchema` is already deprecated and got removed on the `next` branch entirely. That's why it's only supported by the older signature (`Machine`).

**Comment by davydof** [comment]
1. what is the reason of deprecation?
2. what is the further solution - how I can check that states are equal schema?

**Comment by Andarist** [comment]
> what is the reason of deprecation?

It's rarely used by people and it feels like it requires typing the same thing (schema) twice - which is redundant. 

> what is the further solution - how I can check that states are equal schema?

Why do you want to make sure that your config matches some predefined schema? Can't the config act as the source of truth? Why do you want to double-check this? With this schema, every change to the structure requires you to change both the schema and the config.

**Comment by davydof** [comment]
> It's rarely used by people and it feels like it requires typing the same thing (schema) twice - which is redundant.

I extract schema to interface so it's not a problem to check it twice

> Why do you want to make sure that your config matches some predefined schema? Can't the config act as the source of truth? Why do you want to double-check this?

Machine can be large so to see all state changes can be hard but if I have some schema checker it helps me to verify that I didn't break anything

> With this schema, every change to the structure requires you to change both the schema and the config.

It's not a problem, it's good because I can see all changes and verify that updates are going correctly

**Comment by Andarist** [comment]
Do you believe that better tooling would help you to achieve the same thing and certainty without this redundancy? For instance, would a visual diff (generated based on our visualizer) between previous and current state config help you? If this generic would never be a thing in XState - do you think that you would actively seek something like this? Or maybe it's more like once you know that this was a thing you are afraid of losing it?

**Comment by davydof** [comment]
We have developed big complex app on xstate + react. We have one big root machine and a lot of nested machines that are spawn on demand and set to root machine context.
When we need to refactor something it's easier to take a look on schema instead of looking on large root machine.
So, yes - it can help to refactor/improve machine

**Comment by Andarist** [comment]
> As well, personally for me: as much as I can add checks with TS types or add tests to verify that business logic works correctly - I do.

Right, this is my own rule of thumb too. It's just not that obvious where the line is with the schema. Even if we consider automatic/unit/e2e tests - sure, they test our code but what is testing the code of those tests? 😉 

It just feels to me that schema is some way of double-checking things, like a test code that tests the test and not the production code. And I somewhat question its usefulness. So I'm trying understand your PoV to figure out if we can offer some alternatives to it

> So probably schema interface can be optional and consumer decide use or ignore it.

Even though we could make it optional - having it there has no zero costs for maintainability and complexity of our types and that complexity is already pretty high. So if we could remove it without losing an important functionality - that would be a net gain for the codebase.
